### DISCUSSION TITLE
MongoDB connector can not select data from read-only node

### DISCUSSION THREAD

**Discussion by alexanderhw** [trinodb/trino#23123] [Q&A] [answered]
When I recently deployed trino in stand-alone mode and connected it to mongodb, I was unable to read data from the mongodb read-only instance node. **The result indicate that the read-only instance node is not master node of mongodb**. The specific error message is：

```txt
2024-08-21T21:58:25.873+0800    INFO    Query-20240821_135825_00018_swwvh-199   org.mongodb.driver.cluster      Waiting for server to become available for operation with ID 125. Remaining time: 30000 ms. Selector: ReadPreferenceServerSelector{readPreference=primary}, topology description: {type=UNKNOWN, servers=[{address=xxxxxxxxx:8635, type=UNKNOWN, state=CONNECTING, exception={com.mongodb.MongoNotPrimaryException: Command failed with error 10107 (NotMaster): 'not master' on server xxxxxxx:8635. The full response is {"operationTime": {"$timestamp": {"t": 1724248705, "i": 10}}, "ok": 0.0, "errmsg": "**not master**", "code": 10107, "codeName": "**NotMaster**", "$clusterTime": {"clusterTime": {"$timestamp": {"t": 1724248705, "i": 10}}, "signature": {"hash": {"$binary": {"base64": "J7HW9Nj7fF7lRpJJiIS/4z2zJP0=", "subType": "00"}}, "keyId": 7378427933884416004}}}}}].
2024-08-21T21:58:25.920+0800    INFO    dispatcher-query-10     io.trino.event.QueryMonitor     TIMELINE: Query 20240821_135825_00018_swwvh :: FAILED (USER_CANCELED) :: elapsed 46ms :: planning 46ms :: waiting 0ms :: scheduling 0ms :: running 0ms :: finishing 0ms :: begin 2024-08-21T21:58:25.872+08:00 :: end 2024-08-21T21:58:25.918+08:00
2024-08-21T21:58:25.930+0800    ERROR   page-buffer-client-callback-3   io.trino.operator.HttpPageBufferClient  Request to delete http://172.31.33.113:8080/v1/task/20240821_135825_00017_swwvh.1.0.0/results/0 failed java.util.concurrent.CancellationException: Task was cancelled.
```

I noticed that the official deployment guide mentioned that trino needs to have write permissions for information schema. Does this mean that mongodb read-only instance nodes cannot connect to trino?

If trino can connect to mongodb read-only instance nodes, do I need to add any configuration? My current mongodb connection config is:

```properties
connector.name=mongodb
mongodb.connection-url=mongodb://<user>:<pwd>@<ip>:<port>/?authSource=admin
```

**basic environment:**
* trino version: 453
* jdk version: zulu22.32.15-ca-jdk22.0.2-linux_x64
* python version: 3.7
* trino deployment mode: single machine

**mongodb cluster environment:**
* node1: primary
* node2: secondary
* relicaSet: replica
* node3: read-only node represent for the real mongo replica

Update:

I have read the source code of `trino-mongodb` connector and noticed that the `MongoSession#getTableMetadata(String schemaName, String tableName) ` function will read configuration `mongodb.schema-collection` to infer the collection field structure. If the schema collection does not defined, trino needs to create this collection and dynamically maintain it, so mongodb's write permission may be required.

So will trino support maintaining similar table structures within the cluster instead of in specific data source instances? In many production scenarios, from the perspective of data security, data analysis is often connected to read-only instances.

**Comment by ErikAhlund** [comment]
I have a very similar situation, where basically I would like to give trino only read access to the data. Since it needs to maintain the _schema document, is there any way this document could exist in a different mongo database (within the same mongo instance)? Then the document would be named something like "_db-name_schema"

**Reply by unkn0wn-kgb** [reply]
Any update on this? This seem to be a big issue for production setups.
Is it possible somehow for trino to store schema-collection somewhere else?
So it would be possible to run queries on Hidden/Delayed/Secondary replica?

Thanks

**Reply by alexanderhw** [reply, marked as answer]
Now we have solved this problem in another way. We have made a second modification based on the source code of trino, and added two methods of external database & memory to store the Mongodb table structure, which is equivalent to managing the schema-collection function separately.

**Reply by tim-pvm** [reply]
Is it available now? What are those parameters called? Can you point to the code/example?

Thank you

**Reply by alexanderhw** [reply]
I can't show you the entire code because it's hosted within the company, but I can show you a few steps to achieve this functionality.
If you need to store the mongodb table structure in another database, such as MySQL, here are a few steps you need to do.

1. Create a Config class to read relevant configurations from your <mongodb-connector>.properties
```java
// create this java file at plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoExtensionConfig.java
public class MongoExtensionConfig 
{
	private String externalDbUrl;
	private String externalDbUsername;
	private String externalDbPassword;

	@io.airlift.configuration.Config("extenral.db.url")
	public MongoExtensionConfig setExternalDbUrl(String externalDbUrl)
	{
		this.externalDbUrl = externalDbUrl;
		return this;
	}
	
	@io.airlift.configuration.Config("extenral.db.username")
	public MongoExtensionConfig setExternalDbUsername(String externalDbUsername)
	{
		this.externalDbUsername = externalDbUsername;
		return this;
	}
	
	@io.airlift.configuration.Config("extenral.db.password")
	public MongoExtensionConfig setExternalDbPassword(String externalDbPassword)
	{
		this.externalDbPassword = externalDdPassword;
		return this;
	}
	
	// getters ...
}
```

You can add those three config item in your mongodb-connector.properties:
```properties
mongodb.connection-url=
connector.name=mongodb
external.db.url=
external.db.username=
external.db.password=
```

2. Define your mongodb table structure maintainer class
```java
// create this java file at plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoMaintainer.java
public class MongoMaintainer
{
	private MongoExtensionConfig config;

	@com.google.Inject
	public MongoMaintainer(MongoExtensionConfig config)
	{
		this.config = requireNonNull(config, "config is null");
	}
	
	public org.bson.Document getTableStructure()
	{
		// read your self-maintained table structure
		Document doc = read();
		return doc;
	}
	
	public Connection getConnection()
	{
		return DriverManager.getConnection(config.getExternalDbUrl(), config.getExternalDbUsername(), config.getExternalDbPassword());
	}
}
```

3. Register you mongodb table structure maintainer and Config into Mongodb plugin
```java
// plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoClientModule.java

public class MongoClientModule
		extends AbstractConfigurationAwareModule
{
	@Override
    public void setup(Binder binder)
    {
    	// register your config class
    	configBinder(binder).bindConfig(MongoExtensionConfig.class);
    	// register your maintainer
    	config.bind(MongoMaintainer.class).in(Scopes.SINGLETON);
    	
    	// ....
    }
}
```

4. Use your maintainer in MongoSession
```java
// plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java
public class MongoSession
{
	@Inject
	private MongoMaintainer maintainer;
	
	// modify getTableMetadata()
	private Document getTableMetadata(String schemaName, String tableName)
            throws TableNotFoundException
   {
   		Document doc = maintainer.getTableStructure();
   		// You can use the guessTableFields function comprehensively to initialize the table structure
   		return doc;
   }
}
```

By the way, the legal Document structure representing a mongo table should be:
```json
{
    "table": "my_collection",
    "fields": [
        {
            "name": "_id",
            "type": "ObjectId",
            "hidden": false
        },
        {
            "name": "projectId",
            "type": "varchar",
            "hidden": false
        },
        {
            "name": "info",
            "type": "array(row("effect" bigint))",
            "hidden": false
        }
    ]
}
```

**Comment by lu-xiaoshuang** [comment]
What about this option? @alexanderhw @ErikTheBerik 

`https://trino.io/docs/current/connector/mongodb.html#mongodb-read-preference`

**Reply by tim-pvm** [reply]
It does not help as trino requires write collection for internal index/fiels resolve structure.
Ideally it would be great if you could provide a separate connection string for the trinoSchema

**Reply by lu-xiaoshuang** [reply]
In replica set mode, readPreference and/or readPreferenceTags probably make trino route low write traffic and high read traffic to primary node and secondary nodes respectively.
